### ISSUE TITLE
Warn about unused Awaitables (and potentially other objects)

### ISSUE THREAD

**Issue by JelleZijlstra** [python/mypy#2499]
This code:

```
async def f() -> None:
    pass
async def g() -> None:
    f()
```

is incorrect because `f()` isn't awaited. This will produce a runtime warning with the right debug settings, but mypy doesn't give any errors.

It would be nice if mypy could catch this, since I think it's a pretty common error in async code. This could be generalized to warning on other types or function return values that shouldn't be ignored (maybe `open`?), similar to `__attribute__((warn_unused_result))` in [GCC](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes).

**Comment by gvanrossum** [comment]
I'm pretty sure we've got another similar issue somewhere, but I can't find it. (About results that should never be discarded -- perhaps uncalled functions?)

I'm not sure how easy this would be to do -- especially not the generalization, which probably would require us to invent some acceptable syntax for marking functions whose result should never be ignored.

**Comment by Daenyth** [comment]
I think skipping generalization and focusing on `Awaitable` first would still catch bugs. Once there's a second case people want to specialize it'll be easier to see what needs to be made general

**Comment by ilevkivskyi** [comment]
@njsmith 
> Is it possible to have a mypy plugin that adds a check like this?

I am not sure writing it as a plugin is the best solution. This will need addition of at least one new plugin hook. I think a better idea is to just add a flag to mypy. This looks like a common scenario. Also the implementation in mypy is very simple. We can add a flag `is_awaited` on `CallExpr` that will be populated during semantic analysis (by enclosing `visit_await`). An then an error can be given in `visit_call_expr` in `checkexpr.py` if the inferred resulting type is `Awaitable`, but the the flag is not set.

**Comment by ilevkivskyi** [comment]
>  Is mypy interested in carrying features like that?

I have heard similar things from other people. We can start with something, and then maybe generalize the flag to be tunable, for example:
* `--enforce-await=none` (default) does nothing
* `--enforce-await=global` (hard to implement) will complain only if a coroutine is never awaited
* `--enforce-await=local` (easy to implement) will complain if a coroutine is not _immediately_ awaited.

But we can start just with a binary flag `--enforce-await` and then iterate on the result later.

**Comment by JukkaL** [comment]
I don't like the idea of just adding another opt-in warning for this. I'd much prefer if we could have something that is enabled by default, even if it doesn't catch all errors. There's a risk that the majority of users would not realize that such an option exists.

What about always rejecting code such as the original example, as the first iteration of this feature:

```py
async def f() -> None:
    pass
async def g() -> None:
    f()  # Should be an error
```

Is there ever a good reason to write `f()` without doing anything with the result? This wouldn't reject things like `x = f()`, only bare `f()`. `x = f()` would likely generate an error already if the code does anything with `x` later on.

We could later warn about things like `"{}".format(should_be_awaited())` -- again, this is something that would hardly ever be right. We could special case some common operations that work with arbitrary objects, such as `str.format`, `%` formatting, `str(x)` and f-strings. We could extend this to check for things like `"%s" % foo.method` as well (i.e., forgot to call a method).

Later, we could consider adding a trio-specific stricter check (possible as a plugin).

**Comment by JukkaL** [comment]
> Would there be any simple, acceptable change to the plugin interface that would make it possible to implement this as a plugin relatively soon?

Small plugin API changes are typically not controversial, so this is a possibility.

> Maybe there could be some way to run multiple hooks in sequence?

That would be one way to do it, but ordering issues would make this a bit awkward, as you point out. Another option would to add a new "check only" hook that is called after the signature hooks have already run and we have a final signature. The hook wouldn't be able to change the signature but it can perform additional checks, and it's guaranteed to run after the signature-related hooks. We could allow multiple plugins to define these hooks for a function. 

> There's no way for the hook to tell whether a given call is the immediate child of an await node. I guess this could be hacked into the Context?

We'd need another attribute in the relevant context named tuple, I guess.

Can you create a new issue for the plugin feature? We can then focus on the non-trio/non-plugin aspects in this issue.

**Comment by njsmith** [comment]
> Can you create a new issue for the plugin feature? We can then focus on the non-trio/non-plugin aspects in this issue.

OK, forked this off into #5650 and replied there.

**Comment by Dreamsorcerer** [comment]
>     * `--enforce-await=local` (easy to implement) will complain if a coroutine is not _immediately_ awaited.

What about a local check, with the added assumption that a function expecting an `Awaitable` is likely to `await` on it? Plus a tunable unknown state when not enough typing information is available.

This would automatically make functions like `asyncio.gather()` work without needing to special case them. I'm imagining that cases where a function expects an `Awaitable` and then doesn't `await` on it are going to be very rare (plus if they are within the user's code they would also highlight an error).

For example:
```
async def do_something():
    ...

async def func_with_await(foo: Awaitable):
    return foo  # Error: foo not awaited

async def unknown_func(foo: Any):
    print(foo)

async def main():
    # These start as awaited=False
    a = do_something()
    b = do_something()
    c = do_something()
    d = do_something()

    await func_with_await(a)  # a.awaited=True
    asyncio.create_task(b)  # b.awaited=True
    await unknown_func(c)  # c.awaited=Unknown
    # Error: d.awaited=False
```

Whenever the awaited state is `False` it should error, and if `True` it should not error. When it is in the unknown state, this should error depending on a tunable flag (similar to the various `Any` warnings that are configurable).

**Comment by Dreamsorcerer** [comment]
Is there likely to be any movement on this? I've been hitting this problem a lot lately, trying to refactor a bunch of sync code to be async and struggling to find every location that needs an `await`.

**Comment by JelleZijlstra** [comment]
As you see, not much is happening.

I did implement this check in Quora's typechecker, pyanalyze (https://github.com/quora/pyanalyze); you can see the relevant tests at https://github.com/quora/pyanalyze/blob/26e622e249d5b101323647a29ede386d4de06f1e/pyanalyze/test_name_check_visitor.py#L1868. It is not nearly as well-tested as mypy, though, but feel free to give it a try.

**Comment by jstasiak** [comment]
I have a generic proof of concept in https://github.com/python/mypy/issues/8951. It should help in this case too. I've meant to make a proper PR out of it for some time now, I'll try to get it done eventually.

**Comment by erictraut** [comment]
We had a similar discussion in [this pyright issue](https://github.com/microsoft/pyright/issues/1317). As others have mentioned in this thread, there are subtleties involved. I ended up implementing a check in pyright that covers the most common error cases (including the one at the top of this thread) while avoiding most false positives.

If someone is interested in implementing a similar check in mypy, I'm happy to share the learnings from pyright.

**Comment by Dreamsorcerer** [comment]
Thanks, I'll take a look at some of those.

Would be nice if we can atleast decide on the correct approach and remove the needs-discussion label though. Is my proposal above (https://github.com/python/mypy/issues/2499#issuecomment-856909583) reasonable? Or any issues with it? I think it would work 99% of the time and seems like a relatively easy approach to implement (although I'm not really familiar with the mypy codebase).

**Comment by AlexWaygood** [comment]
There has recently been movement on this in the form of #12279

**Comment by Dreamsorcerer** [comment]
Looks like a good start. Judging by the examples, I assume this just checks that the return value is assigned to a variable. Which will catch the simplest mistakes like:
```
def f():
    db.insert(...)
```

But, it won't yet check if the return value is actually awaited.

**Comment by Daenyth** [comment]
What about this design:
- always warn when an awaitable is not awaited
- provide a `cast`-like method (implemented as `identity`) to signal to mypy "no, really, I don't want to await this"

**Comment by graingert** [comment]
anyio has the concept of DeprecatedAwaitables that warn when you await them - they are list, float and Noneish subclasses otherwise

**Comment by jhance** [comment]
I think trying to find whether all possible control flows await an awaitable is difficult for a type checker. How does it know if I pass it to a function whether that function will await it or not?

**Comment by Dreamsorcerer** [comment]
> I think trying to find whether all possible control flows await an awaitable is difficult for a type checker. How does it know if I pass it to a function whether that function will await it or not?
> 
> The statement that passing it to a function is likely to await it is not true. For example appending to a list

How frequently do you think a function will explicitly declare the type as `Awaitable` and then not await on it (either directly or through another function)? My assumption is that this will be very rare in real world use cases, likewise a typesafe function that doesn't accept an `Awaitable` (e.g. `object`) is very unlikely to await on it. If this works in real world use cases 99% of the time, then it's definitely worth putting in.

If a function accepts `Any` then we have no idea if it will await or not, so having a strictness setting to warn or not in these cases is helpful. We can default to not warning to avoid false positives, but people looking for some stricter/safer can enable warnings in this case (see the different examples in my comment).

**Comment by graingert** [comment]
https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-awaitable-return-value-is-used-unused-awaitable

**Comment by graingert** [comment]
closed in https://github.com/python/mypy/pull/12279 ?

see docs here https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-awaitable-return-value-is-used-unused-awaitable

**Comment by Dreamsorcerer** [comment]
That just handles unused awaitable expressions. Something as simple as `a = foo()` won't trigger the warning.

I think we still need to have something like https://github.com/python/mypy/issues/2499#issuecomment-856909583 to cover a lot more cases. It's a bit more complex as you need to track the awaited state through a function in order to see if it was awaited by the end, but would catch a lot more mistakes.

**Comment by jstasiak** [comment]
`a = foo()` where `a` remains unused should be covered by flake8 or other linters already under the unused variable umbrella (granted, maybe it'd be nice to have it in Mypy as well)

**Comment by Dreamsorcerer** [comment]
> `a = foo()` where `a` remains unused should be covered by flake8 or other linters already under the unused variable umbrella (granted, maybe it'd be nice to have it in Mypy as well)

That only works if the variable is unused. You can then do `print(a)` or similar and flake8 will no longer catch it as unused and mypy will not catch it as unawaited.

**Comment by tback** [comment]
Can mypy be configured to require type hints for unawaited assigns? 

```
async def do_something(): ...

async def main():
    # These start as awaited=False
    a = do_something()  # ERROR
    b : Awaitable = do_something()  # OK
    c = await do_something()  # OK
```

**Comment by Dreamsorcerer** [comment]
That is not an unusual thing to do though, so not sure how many people would find such an option useful.

**Comment by tback** [comment]
My reasoning behind this request is to make delayed calls an active decision and to prevent errors from forgotten awaits.
I'd assume that this would provide a robust solution for a couple of "simple" use cases while also being kind of easy to implement.

**Comment by blast-hardcheese** [comment]
I ran into an issue recently with the following construct:

```python
async def a() -> bool:
  return False

if a():
  pass
```

Because `a` was truthy, it would always pass. The way I interpret the conversation thus far is to move towards disambiguation, guiding the user to codify their intent so typecheckers can pick up the slack `MustUse[...]` is an interesting idea (and fits my aesthetic) but I agree with the author that that's actually the default expectation in most cases.

---

I think a sensible nudge here would be `Awaitable[...]` values requiring an explicit round-trip through a primitive type before being used in a primitive way. In my thinking currently, either `bool` or `str`. I don't think others apply.


### bool
In my example above, if I truly wanted truthy behavior, I would change it to `if bool(a()):`.

Interestingly, the `None` case is already handled:

```python
def a() -> None | Awaitable[...]:
  ...

if await a():
  pass
```

relying on `await None` being a no-op is sensible.

### str

This would also sensibly apply to string interpolation. The following could be disambiguated by suggesting the user explicitly choose `str(a())` or `repr(a())` depending on what they are trying to accomplish.

```python
f"The value is currently {a()}."
```

This would help with code reviews as well, where more seasoned developers would see the out-of-place `str(...)` or `bool(...)` and start prodding.

---

I think broadly, if you're using a typechecker in Python at all you're opting in to removing classes of bugs.

What we are trying to do here is just a formalization of intent, and shouldn't remove expressivity from the language.

I could see pushback here being adding too much noise to existing programs, but I'd assert that the vast majority of cases where this would flag on existing code are actually already bugs (or debugging statements which can be patched by rote)

**Comment by hauntsaninja** [comment]
Since people at work were talking about this... Mypy already supports a version of this; it will complain about OP's case out of the box. For others, you might need `--enable-error-code unused-awaitable`. Also `--enable-error-code truthy-bool` can catch some similar issues like the one from @blast-hardcheese .
